Add playback to Drake visualizer

SeanCurtis-TRI opened the pull request:
1) The `SignalLog` class handles the basic logic for caching the time-dependent
data.
2) `SignalLogger` and `DrakeVisualizer` both own an instance of SignalLog
3) `DrakeVisualizer` conditionally populates and replays the signal log

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/5152)
<!-- Reviewable:end -->

SeanCurtis-TRI commented:
+@russtedrake for feature review +@sherm1 for platform review

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-KcsGspEHbSXu2tdBi8i:b-7coyi9)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
This seem like poor separation-of-concerns if we have to frob the DrakeVisualizer system in order to do this.  I can take a look much later in the day, but I'll ask first: did this get a design review yet?

sherm1 commented:
This was a mutual concoction by @seancurtis-tri, @russtedrake, and me, and seemed better than the earlier alternatives we came up with. Playback as a feature of the DrakeVisualizer system seems reasonable to me, what do you think would be better?

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-KcsKwnzosTUffR66Ipz:bw893g2)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
On first glance, all of this code could go into a separate class with essentially no downside?

SeanCurtis-TRI commented:
The biggest advantage of this design is any code that currently includes a `DrakeVisualizer` in its `Diagram` would only need to add a single boolean to its instantiation to enable playback.  (And make a single call to that instance to execute playback).  As such, it is "minimally" invasive.  I had created an earlier PR in which the caching and playback happened in a different system, yielding a much more weighty solution.

---

Review status: 0 of 8 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-KcsLr9BWHa-s2b7lePY:b-r1pfvb)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Sure.  But you can still do that even with this code moved into another class, right?

SeanCurtis-TRI commented:
So, if I understand what you're saying, you don't have a problem with the API on `DrakeVisualizer` you would just want this implementation encapsulated somewhere else?

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-KcsNSmS0xvsqq_p4Qu8:bnb03gz)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
It would at least be substantially better that way, and then we can haggle about the DrakeVisualizer system sugar once that's in.  (For example, I suspect that `enable_recording(bool)` is superior to a constructor argument, and that a getter on DV for the helper class is superior to putting a "Playback" methods directly onto DV.)

sherm1 commented:
Hmmm ... I like that @seancurtis-tri pulled out the guts of SignalLogger into a separate class but I'm not seeing the advantage of a further separation. Is there an abstraction you have in mind Jeremy that would be represented by the additional class?

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-KcsQxq429-NW4wyGdl4:b-impqsc)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
But then we'd have two-way coupling.  The DV would have to know about caching player so it can provide the data that gets cached.  The caching player would have to know about the DV in order to translate cached state into LCM messages.  

---

Review status: 0 of 9 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-KcsR5dD2B2LTu-HB8yT:bbv75xk)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Well that's not true (the bi-di coupling).

However, (1) I have no time to explain and (2) actually, I think this is not a problem that `automotive` cares about since we are not going to use DrakeVisualizer System anymore as soon as we can possibly manage it.  (Because it requires RBT, and we hate RBT with the fiery passion of a trillion dying suns.)

So I guess you can go ahead with this, and I'll fix it later if I end up still using DrakeVisualizer System longer that I predicted.

SeanCurtis-TRI commented:
So, back to you @russtedrake for feature review.

---

Review status: 0 of 10 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-KcsZtzMJm6GXMv3EXrs:b-1x5hsg)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
hmmm, i dont understand. we have playback for free from lcm logger.

liangfok commented:
Review status: 0 of 10 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/multibody/rigid_body_plant/drake_visualizer.cc, line 81 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcslH2dmkjsfgC09lhd:-KcslH2dmkjsfgC09lhe:b-c9ipe3) ([raw file](https://github.com/robotlocomotion/drake/blob/7b38b6c44869e69a53809a2a24057a64e879b2d3/drake/multibody/rigid_body_plant/drake_visualizer.cc#L81)):*
> ```cpp
>     data.set_value(trajectory.value(sim_time));
> 
>     // Evaluate at time
> ```

BTW, sentence fragment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152)*
<!-- Sent from Reviewable.io -->

siyuanfeng-tri commented:
am i missing something here? if you start a lcm-logger with your simulation, it will log all the lcm messages, which includes the drawing commands in channel DRAKE_VIEWER_DRAW. if you want to replay the simulation visuals, start the drake_visualizer, and lcm-logplayer-gui with the log file name, you would have all the visualization.
if you want an in memory log of the state vector, that's a different story. 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-KcssiW4ilSCYvRSg7x6:b-fd1xtq)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
@siyuanfeng-tri -- the point is to play back slow simulations at real-time rate.  this is really important for "visual debugging" of dynamics.  you can't do that from an lcm log.

jwnimmer-tri commented:
The LCM log player can't govern its playback rate by the in-band timestamps?  That's absolutely broken, and seems worth fixing, whether or not this PR also goes in.

RussTedrake commented:
There are multiple timestamps.  The clock and the simtime.  lcm, and therefore lcm-logger, doesn't have any protocol/standard that it would implement for pulling simtime out of the message and playing back on that.  It *does* play back at the same rate things were recorded.

jwnimmer-tri commented:
Right.  That's broken.  We should fix it (or write a new tool) that allows us to use the in-band timestamps.  And if LCM can't do that, we should stop using LCM.

sammy-tri commented:
>  It does play back at the same rate things were recorded.

That seems to imply that there is a timestamp associated with the logged packets in the recording?  Even if it's only log-relative or something?  (and it also implies it could play back at a multiple of the recording rate)

RussTedrake commented:
you can play it back at a multiple.  but our sim doesn't play at a multiple of real time.  if it's slow, then it's not locked to the wall clock in any way.  but even more important is the variable-step solver case ... where you simulate carefully, and want to watch a physically realtime simulation after the fact.

rdeits commented:
@jwnimmer-tri @sammy-tri every time the LCM logger receives a message it stores that message and the wall-clock time at which it was received. As long at the message itself contains a sim-time timestamp, it would be very easy to produce a tool that, for example, processes an LCM log to extract a channel's messages and replace their playback times with their sim-time timestamps, thus enabling real-time playback.

RussTedrake commented:
a few small comments.  but <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> .  this will be very useful.  thanks!

---

Reviewed 10 of 10 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/drake_visualizer.cc, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcu4qWWXf_WWfJv7BRl:-Kcu4qWWXf_WWfJv7BRm:b-b7ioc) ([raw file](https://github.com/robotlocomotion/drake/blob/7b38b6c44869e69a53809a2a24057a64e879b2d3/drake/multibody/rigid_body_plant/drake_visualizer.cc#L34)):*
> ```cpp
> 
> void DrakeVisualizer::ReplayCachedSimulation() {
>   if (has_playback_) {
> ```

perhaps alert with a helpful message if replay is called without the playback?

---

*[drake/multibody/rigid_body_plant/drake_visualizer.cc, line 78 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcu582YhOIRvD5sMfHh:-Kcu582ZWqVFJ9rr7LFS:b-3bmuyd) ([raw file](https://github.com/robotlocomotion/drake/blob/7b38b6c44869e69a53809a2a24057a64e879b2d3/drake/multibody/rigid_body_plant/drake_visualizer.cc#L78)):*
> ```cpp
>   TimePoint prev_time = Clock::now();
>   BasicVector<double> data(log_->get_input_size());
>   while (sim_time <= kMaxTime) {
> ```

i've found it very important/useful to make sure you explicitly publish the last frame of the simulation.  often we're checking something (even visually) at the final time.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-Kcu5UDBl64JNwJaA8xO:b4f9iiv)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Comments addressed.  PTAL

---

Review status: 9 of 10 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/drake_visualizer.cc, line 34 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcu4qWWXf_WWfJv7BRl:-KcxED6-T0MZAyV0agtZ:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/7b38b6c44869e69a53809a2a24057a64e879b2d3/drake/multibody/rigid_body_plant/drake_visualizer.cc#L34)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

perhaps alert with a helpful message if replay is called without the playback?
</blockquote></details>

Done

---

*[drake/multibody/rigid_body_plant/drake_visualizer.cc, line 78 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcu582YhOIRvD5sMfHh:-KcxDvs_B8G3Ssa9N_Yq:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/7b38b6c44869e69a53809a2a24057a64e879b2d3/drake/multibody/rigid_body_plant/drake_visualizer.cc#L78)):*
<details><summary><i>Previously, RussTedrake (Russ Tedrake) wrote…</i></summary><blockquote>

i've found it very important/useful to make sure you explicitly publish the last frame of the simulation.  often we're checking something (even visually) at the final time.
</blockquote></details>

Done

---

*[drake/multibody/rigid_body_plant/drake_visualizer.cc, line 81 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcslH2dmkjsfgC09lhd:-KcxEBmlHLNUu-Oj71QD:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/7b38b6c44869e69a53809a2a24057a64e879b2d3/drake/multibody/rigid_body_plant/drake_visualizer.cc#L81)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, sentence fragment.
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-KcxEfbOewgLLEDVGk3L:ba8prf5)*
<!-- Sent from Reviewable.io -->

liangfok commented:
One problem I can see with the current design of including the logging & replay logic in `DrakeVisualizer` is it requires a `DrakeVisualizer` to be instantiated to be able to support simulation playback. `DrakeVisualizer` is not available in minimal builds since LCM is not present. I would be nice to have the ability to do playback even on applications that do not make use of `DrakeVisualizer` or on platforms without LCM.

---

Review status: 9 of 10 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/drake_visualizer.cc, line 36 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxJ1GQimF16SZiTFKG:-KcxJ1GQimF16SZiTFKH:b-tr9tbx) ([raw file](https://github.com/robotlocomotion/drake/blob/c96013f664daa08e52fc67198e4b176a6bde55b2/drake/multibody/rigid_body_plant/drake_visualizer.cc#L36)):*
> ```cpp
> void DrakeVisualizer::ReplayCachedSimulation() {
>   if (has_playback_) {
>     // Build piecewise polynomial
> ```

BTW, missing period at end of sentence.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-KcxJlzYn4Eq6ghNRbmc:bh1u3c3)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Your point is valid.  However, this feature doesn't aspire to be the grand solution for all simulation playback.  In fact, the previously discussed solution -- using the LcmLogger and playback -- is likewise unsuitable for the scenario you describe.  However, this doesn't aspire to be a universal solution.  It is a marginal amount of code to serve a limited purpose.  It is *ideal* for those scenarios where I construct a simulation scenario that has some limited simulation time domain that I want to observe directly in real time.  If I'm already inserting a `DrakeVisualizer` system into my `Diagram` then there are better than even odds that this applies to me.  If I've got something larger scale than that, it may not apply to me.

I can provide a longer laundry list of what this code can't accomplish:
1) I'm playing back at 60 Hz, but am *saving* data at the simulation rate.
2) It doesn't include contact information (that is processed and broadcast by a different system).
3) It's probably impractical for a simulation that takes hours to create the amount of cached data I'm interested in (but I nevertheless have to have a drake visualizer running the entire time.)
etc.

---

Review status: 9 of 10 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/drake_visualizer.cc, line 36 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxJ1GQimF16SZiTFKG:-KcxLkvA7RMIc9Pwvrxs:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/c96013f664daa08e52fc67198e4b176a6bde55b2/drake/multibody/rigid_body_plant/drake_visualizer.cc#L36)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, missing period at end of sentence.
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-KcxLmDdXgC-5i4UMVsJ:b-ld7sf7)*
<!-- Sent from Reviewable.io -->

liangfok commented:
All very good points!

---

Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-KcxNQ-We66-M0yZtYLr:b-62bxb)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
So, it's not a universal hammer.  It's more like a cobbler's hammer for driving in a very specific kind of brad through leather.  Sucks for building a house, but amazingly effective at resoling shoes.  :)

---

Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-KcxVfcBHdeZc-i84pcC:blff3re)*
<!-- Sent from Reviewable.io -->

liangfok commented:
And the design employed by this PR does not prevent a future design where the recording and playback is done in a manner that's independent of LCM / DrakeVisualizer, so all is good.

---

Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-KcxWQBmmVC0bnd8oXo6:btmbihl)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
I agree.

---

Review status: 9 of 10 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-KcxWw4hII4RvkCVdS1s:btd51a6)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I suppose it does result in the need to clean up some code shrapnel (like the inclusion of `piecewise_polynomial` and `signal_log` within `DrakeVisualizer` once the logging & playback feature is done by a system that is separate from the `DrakeVisualizer`. Oh well, it's not a big deal.

---

Reviewed 9 of 10 files at r1, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 23 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/drake_visualizer.cc, line 22 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxZ9iDH_Lvz_0ZsGTH:-KcxZ9iDH_Lvz_0ZsGTI:b-93g9u6) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.cc#L22)):*
> ```cpp
>       load_message_(CreateLoadMessage(tree)),
>       draw_message_translator_(tree),
>       has_playback_(record_for_playback) {
> ```

BTW, interesting mix of variable names and semantics. Isn't it true that `record_for_playback` is true while `has_playback_` is actually false when the simulation hasn't started yet?

---

*[drake/multibody/rigid_body_plant/drake_visualizer.cc, line 66 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxXXaDdjvKhPngpDpM:-KcxXXaDdjvKhPngpDpN:bl85d0d) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.cc#L66)):*
> ```cpp
>   } else {
>     drake::log()->warn(
>         "DrakeVisualizer::ReplayCachedSimulation called on instance that wasn't"
> ```

BTW, consider adding "()" after "ReplayCachedSimulation" so readers know it's a method.

---

*[drake/multibody/rigid_body_plant/drake_visualizer.cc, line 67 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxXktB3_GFyO1Pc0Li:-KcxXktB3_GFyO1Pc0Lj:b-k83byn) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.cc#L67)):*
> ```cpp
>     drake::log()->warn(
>         "DrakeVisualizer::ReplayCachedSimulation called on instance that wasn't"
>         " initialized to record.");
> ```

BTW, consider including a suggestion saying "Next time, please construct `DrakeVisualizer` with recording enabled!".

---

*[drake/multibody/rigid_body_plant/drake_visualizer.cc, line 80 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxY7R_43axxO7PWFDw:-KcxY7R_43axxO7PWFDx:b-630w1c) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.cc#L80)):*
> ```cpp
>   // Target frame length at 60 Hz playback rate.
>   const double kFrameLength = 1 / 60.0;
>   double sim_time = 0;
> ```

BTW, could sim_time start at a non-zero value?

---

*[drake/multibody/rigid_body_plant/drake_visualizer.cc, line 88 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxYivVDU5WGhjxpCJv:-KcxYivVDU5WGhjxpCJw:b-kmsjvd) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.cc#L88)):*
> ```cpp
>     // Translates the input vector into an array of bytes representing an LCM
>     // message.
>     std::vector<uint8_t> message_bytes;
> ```

BTW, any concern about the need to allocate deallocate this each cycle of this loop?

---

*[drake/multibody/rigid_body_plant/drake_visualizer.cc, line 103 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxYZhp1IIllLRnD1Vq:-KcxYZhp1IIllLRnD1Vr:b-uc5az7) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.cc#L103)):*
> ```cpp
>   }
> 
>   // Final evaluation is at the final time stamp, guaranteeing the final state.
> ```

BTW, consider adding "is visualized" after "state".

---

*[drake/multibody/rigid_body_plant/drake_visualizer.h, line 43 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcxd4kZ5jz4mObmCZAB:-Kcxd4kZ5jz4mObmCZAC:b-pbj8ry) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.h#L43)):*
> ```objc
>  * "DRAKE_VIEWER_DRAW".
>  *
>  * The visualizer has an option that allows it to save the state it dispatches
> ```

BTW, consider replacing "allows it to" with "makes it".

---

*[drake/multibody/rigid_body_plant/drake_visualizer.h, line 44 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcxa-BJhpA1uvkHP5gf:-KcxZTyciqu2kGzyBujt:b-91clbo) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.h#L44)):*
> ```objc
>  *
>  * The visualizer has an option that allows it to save the state it dispatches
>  * for drawing and then replay it at realtime (capped at 60 Hz).  This is useful
> ```

BTW, "at realtime" is a fairly loaded and thus easily misunderstood term. Consider replacing it with "at wall clock time".

---

*[drake/multibody/rigid_body_plant/drake_visualizer.h, line 45 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcx_LeglZYTfY-Vbfcw:-Kcx_LeglZYTfY-Vbfcx:b-1tnfvl) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.h#L45)):*
> ```objc
>  * The visualizer has an option that allows it to save the state it dispatches
>  * for drawing and then replay it at realtime (capped at 60 Hz).  This is useful
>  * for immediate review of slow simulations to gain an intuitive understanding
> ```

BTW, it's also useful for fast simulations that happen so quickly that it is unobservable by the naked eye right?

---

*[drake/multibody/rigid_body_plant/drake_visualizer.h, line 95 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxaBUN_rEPVqGEwlh2:-KcxaBUN_rEPVqGEwlh3:b-x0kn6l) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.h#L95)):*
> ```objc
>    * not been configured to record/playback, no work is done.
>    */
>   void ReplayCachedSimulation();
> ```

BTW, consider adding a parameter that allows the user to specify a desired reply speed. For example, wouldn't it be useful to allow the replay to run both slower and faster than wall clock​ time?

---

*[drake/multibody/rigid_body_plant/drake_visualizer.h, line 136 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxavTi1B6yy_UdPuAU:-KcxavTi1B6yy_UdPuAV:b-aw4m2a) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.h#L136)):*
> ```objc
>   // If true, DrakeVisualizer records the data it "publishes", making it
>   // available for playback.
>   bool has_playback_{false};
> ```

BTW, consider removing this member variable and simply checking whether `log_ == nullptr`.

---

*[drake/systems/primitives/BUILD, line 168 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcxb9rrKS-maM6WuGsf:-Kcxb9rrKS-maM6WuGsg:bcdzq72) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/systems/primitives/BUILD#L168)):*
> ```Python
>     srcs = ["signal_log.cc"],
>     hdrs = ["signal_log.h"],
>     linkstatic = 1,
> ```

This line might no longer be needed after #5157. Pinging @david-german-tri for confirmation.

---

*[drake/systems/primitives/signal_log.cc, line 21 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxcNOmkhW9AfRQPvxN:-KcxcNOmkhW9AfRQPvxO:b-suut28) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/systems/primitives/signal_log.cc#L21)):*
> ```cpp
> void SignalLog<T>::AddData(T time, VectorX<T> sample) {
>   if (num_samples_ == 0 || time >= sample_times_(num_samples_ - 1))
>     ++num_samples_;
> ```

BTW, should there be a check here for overflow of the `int` data type?

---

*[drake/systems/primitives/signal_log.h, line 10 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxbjsbbKX56p8uDG7I:-KcxbjsbbKX56p8uDG7J:bbws56o) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/systems/primitives/signal_log.h#L10)):*
> ```objc
> 
> /**
>  This class serves as an in-memory cache of the time-dependent vector values.
> ```

BTW, consider deleting "the".

---

*[drake/systems/primitives/signal_log.h, line 19 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcxbo8JDD9q3podUdvh:-Kcxbo8JDD9q3podUdvi:bdyc9s9) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/systems/primitives/signal_log.h#L19)):*
> ```objc
>   DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(SignalLog)
> 
>   /** Construct the signal log.
> ```

BTW, "Construct" should be "Constructs" per style guide.

---

*[drake/systems/primitives/signal_log.h, line 20 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxbsA5dxlyAUJ_LFXE:-KcxbsA5dxlyAUJ_LFXF:bv8y1yh) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/systems/primitives/signal_log.h#L20)):*
> ```objc
> 
>   /** Construct the signal log.
>    @param input_size Dimension of the per-time step data set..
> ```

BTW, extra period.

---

*[drake/systems/primitives/signal_log.h, line 26 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcxc-4CtbJYv0ZMEJi0:-Kcxc-4Duo__PotZEAWE:badbeco) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/systems/primitives/signal_log.h#L26)):*
> ```objc
>   explicit SignalLog(int input_size, int batch_allocation_size = 1000);
> 
>   /** Access the (simulation) time of the logged data. */
> ```

BTW, consider changing "time" to "times".

---

*[drake/systems/primitives/signal_log.h, line 26 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcxc4MEojq4dIUFZz4m:-Kcxc4MEojq4dIUFZz4n:b-h88ju2) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/systems/primitives/signal_log.h#L26)):*
> ```objc
>   explicit SignalLog(int input_size, int batch_allocation_size = 1000);
> 
>   /** Access the (simulation) time of the logged data. */
> ```

BTW, "Access" should be "Accesses" per style guide.

---

*[drake/systems/primitives/signal_log.h, line 31 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcxc8we5VRq6PMUzzMA:-Kcxc8we5VRq6PMUzzMB:b-h88ju2) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/systems/primitives/signal_log.h#L31)):*
> ```objc
>   }
> 
>   /** Access the logged data. */
> ```

BTW, "Access" should be "Accesses" per style guide.

---

*[drake/systems/primitives/signal_log.h, line 37 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxcBZxCSI5FS2-r7QH:-KcxcBZxCSI5FS2-r7QI:b5hslyt) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/systems/primitives/signal_log.h#L37)):*
> ```objc
>   }
> 
>   /** Add a `sample` to the data set with the associated `time` value.
> ```

BTW, "Add" should be "Adds" per style guide.

---

*[drake/systems/primitives/signal_log.h, line 45 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcxcc2bsTA5FdQulLw1:-Kcxcc2bsTA5FdQulLw2:b-d8kd93) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/systems/primitives/signal_log.h#L45)):*
> ```objc
> 
>   /** Reports the size of the log's input vector. */
>   int64_t get_input_size() const { return data_.rows(); }
> ```

BTW, does `data_.rows()` actually return an `int64_t`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-KcxWtqM2ONSBOh5yuDa:b-4ifvui)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
For the record, a design that used a subclass of DrakeLcmInterface (that proxied to real-LCM but snarfed up the messages) would likely be shorter, more elegant, and more powerful than the design proposed in this PR.

SeanCurtis-TRI commented:
Addressed reviewer comments.  PTAL

---

Review status: 5 of 10 files reviewed at latest revision, 23 unresolved discussions.

---

*[drake/multibody/rigid_body_plant/drake_visualizer.cc, line 22 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxZ9iDH_Lvz_0ZsGTH:-Kd1NJMazjx1grFYzeiF:b1ejjdv) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.cc#L22)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, interesting mix of variable names and semantics. Isn't it true that `record_for_playback` is true while `has_playback_` is actually false when the simulation hasn't started yet?
</blockquote></details>

Done `has_playback_` has changed to `playback_enabled_`

---

*[drake/multibody/rigid_body_plant/drake_visualizer.cc, line 66 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxXXaDdjvKhPngpDpM:-Kd1NpIm6EjZx3utzwPw:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.cc#L66)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider adding "()" after "ReplayCachedSimulation" so readers know it's a method.
</blockquote></details>

Done

---

*[drake/multibody/rigid_body_plant/drake_visualizer.cc, line 67 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxXktB3_GFyO1Pc0Li:-Kd1O3Snsks_6JzxcaHL:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.cc#L67)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider including a suggestion saying "Next time, please construct `DrakeVisualizer` with recording enabled!".
</blockquote></details>

Done

---

*[drake/multibody/rigid_body_plant/drake_visualizer.cc, line 80 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxY7R_43axxO7PWFDw:-Kd1O4IrCEgo0yRvLyfG:b7u2ew8) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.cc#L80)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, could sim_time start at a non-zero value?
</blockquote></details>

BTW See the todo comment in the header file.  There are a number of refinements that could be applied to this algorithm.  For now, we're just pushing the simplest.

---

*[drake/multibody/rigid_body_plant/drake_visualizer.cc, line 88 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxYivVDU5WGhjxpCJv:-Kd1OSVJURMZqOLm-EM9:bc0kkxc) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.cc#L88)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, any concern about the need to allocate deallocate this each cycle of this loop?
</blockquote></details>

BTW Only if it turns out we need to replay at some ridiculously high playback rate.  Certainly, 60 Hz is not a problem.  And it keeps the book-keeping simple. 

As I look at the code that is responsible for creating this message, I assume the call to `RigidBodyTree::doKinematics` will dominate over the call to lcm_message_bytes->resize().  That said, it seems reasonable that it *could* be persisted.

---

*[drake/multibody/rigid_body_plant/drake_visualizer.cc, line 103 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxYZhp1IIllLRnD1Vq:-Kd1QchqfEq2ZqSHOqjw:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.cc#L103)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider adding "is visualized" after "state".
</blockquote></details>

Done

---

*[drake/multibody/rigid_body_plant/drake_visualizer.h, line 43 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcxd4kZ5jz4mObmCZAB:-Kd1QqxNGTAIcxnW5wpg:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.h#L43)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider replacing "allows it to" with "makes it".
</blockquote></details>

Done

---

*[drake/multibody/rigid_body_plant/drake_visualizer.h, line 44 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcxa-BJhpA1uvkHP5gf:-Kd1QtbsegMvHUgNcNj5:b-wt4vrb) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.h#L44)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, "at realtime" is a fairly loaded and thus easily misunderstood term. Consider replacing it with "at wall clock time".
</blockquote></details>

Done "realtime" is certainly an overloaded term.  

---

*[drake/multibody/rigid_body_plant/drake_visualizer.h, line 45 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcx_LeglZYTfY-Vbfcw:-Kd1RFe7Sakv4604lL74:b-425hzw) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.h#L45)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, it's also useful for fast simulations that happen so quickly that it is unobservable by the naked eye right?
</blockquote></details>

BTW I hadn't thought of that.  But yes.

---

*[drake/multibody/rigid_body_plant/drake_visualizer.h, line 95 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxaBUN_rEPVqGEwlh2:-Kd1Ribgf6afRu0U094c:bvddhex) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.h#L95)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider adding a parameter that allows the user to specify a desired reply speed. For example, wouldn't it be useful to allow the replay to run both slower and faster than wall clock​ time?
</blockquote></details>

Extended the TODO on optional features.

---

*[drake/multibody/rigid_body_plant/drake_visualizer.h, line 136 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxavTi1B6yy_UdPuAU:-Kd1RtBQRGLjW56PTc_9:b-xceqyt) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/multibody/rigid_body_plant/drake_visualizer.h#L136)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider removing this member variable and simply checking whether `log_ == nullptr`.
</blockquote></details>

BTW  Excellent point.  This was an artifact left over from the previous incarnation.

---

*[drake/systems/primitives/BUILD, line 168 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcxb9rrKS-maM6WuGsf:-Kd1STyWAW8unGKuIkDj:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/systems/primitives/BUILD#L168)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

This line might no longer be needed after #5157. Pinging @david-german-tri for confirmation.
</blockquote></details>

Done

---

*[drake/systems/primitives/signal_log.cc, line 21 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxcNOmkhW9AfRQPvxN:-Kd1TXGPwHl7TY5rIKDs:b5u099v) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/systems/primitives/signal_log.cc#L21)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, should there be a check here for overflow of the `int` data type?
</blockquote></details>

BTW I believe the problem is alleviated by promoting the counter to 64-bits.

---

*[drake/systems/primitives/signal_log.h, line 10 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxbjsbbKX56p8uDG7I:-Kd1U3SMgX0VZ2Gta-7i:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/systems/primitives/signal_log.h#L10)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider deleting "the".
</blockquote></details>

Done

---

*[drake/systems/primitives/signal_log.h, line 19 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcxbo8JDD9q3podUdvh:-Kd1UOEMRxgSjUhf9RIZ:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/systems/primitives/signal_log.h#L19)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, "Construct" should be "Constructs" per style guide.
</blockquote></details>

Done

---

*[drake/systems/primitives/signal_log.h, line 20 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxbsA5dxlyAUJ_LFXE:-Kd1UOmoAHcwktVmnQQs:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/systems/primitives/signal_log.h#L20)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, extra period.
</blockquote></details>

Done

---

*[drake/systems/primitives/signal_log.h, line 26 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcxc-4CtbJYv0ZMEJi0:-Kd1UY9LCDUKEMeb2h80:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/systems/primitives/signal_log.h#L26)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, consider changing "time" to "times".
</blockquote></details>

Done

---

*[drake/systems/primitives/signal_log.h, line 26 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcxc4MEojq4dIUFZz4m:-Kd1UZEhXyloZ0YFnMUj:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/systems/primitives/signal_log.h#L26)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, "Access" should be "Accesses" per style guide.
</blockquote></details>

Done

---

*[drake/systems/primitives/signal_log.h, line 31 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcxc8we5VRq6PMUzzMA:-Kd1U_3ezF71lXFOqmra:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/systems/primitives/signal_log.h#L31)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, "Access" should be "Accesses" per style guide.
</blockquote></details>

Done

---

*[drake/systems/primitives/signal_log.h, line 37 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-KcxcBZxCSI5FS2-r7QH:-Kd1Ua1XgoUM2i4uTUyB:bb6un1j) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/systems/primitives/signal_log.h#L37)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, "Add" should be "Adds" per style guide.
</blockquote></details>

Done

---

*[drake/systems/primitives/signal_log.h, line 45 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-Kcxcc2bsTA5FdQulLw1:-Kd1Ud1RCgF7OyJPQbQ2:b-i5z45g) ([raw file](https://github.com/robotlocomotion/drake/blob/5f4d11c568aa20b86bbd0043a7d9ce07f3368e0e/drake/systems/primitives/signal_log.h#L45)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary><blockquote>

BTW, does `data_.rows()` actually return an `int64_t`?
</blockquote></details>

BTW In a hazy way, yes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-Kd1MHpxSFMetRQ_PX_T:b-ep3l06)*
<!-- Sent from Reviewable.io -->

liangfok commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-Kd1kYNiCzy3SQ6dd1zM:bnfp4nl)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
-@sherm1 

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-Kd2-SaW6bQcSF4O3CBi:blu3ide)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
@russtedrake There are two blocking discussions. Also, do you feel a platform review is still necessary?

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-Kd2-neQ96bnTboXpt5F:b-ochs17)*
<!-- Sent from Reviewable.io -->

liangfok commented:
+@liangfok

---

Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-Kd375LtGkOi7K2ta-n_:b-2u70ip)*
<!-- Sent from Reviewable.io -->

RussTedrake commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> -- i can count as platform review.

---

Reviewed 5 of 5 files at r4.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-Kd3W7qMeo1xJkgdUQW3:bnntjce)*
<!-- Sent from Reviewable.io -->

SeanCurtis-TRI commented:
Great, thanks

---

Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/5152#-:-Kd6YsrrPaggVGzS4B2J:b-4so2ez)*
<!-- Sent from Reviewable.io -->
